lib/src/components/NxFileUpload/NxMultiFileUpload/NxMultiFileUpload.tsx

```diff
@@ -0,0 +1,188 @@
+  const {
+        className: classNameProp,
+        onChange: onChangeProp,
+        files,
+        isRequired,
+        isPristine,
+        id,
+        disabled,
+        ...attrs
+      } = props,
+      isFileSelected = !!files?.length,
+      { showValidationErrors: formShowValidationErrors } = useContext(FormAriaContext),
+      showValidationErrors = formShowValidationErrors || !isPristine,
+      showError = isRequired && showValidationErrors && !isFileSelected,
+      className = classnames('nx-multi-file-upload', classNameProp),
+      noFileMessageClassName = classnames('nx-multi-file-upload__no-file-message', {
+        'nx-multi-file-upload__no-file-message--invalid': showError
+      }),
+      inputRef = useRef<HTMLInputElement>(null),
+      inputId = useUniqueId('nx-multi-file-upload-input', id),
+      selectedFilesContainerRef = useRef<HTMLDivElement>(null),
+      validationErrorId = useUniqueId('nx-multi-file-upload-validation-error'),
+      [previousFiles, setPreviousFiles] = useState<FileList | null>(files);
+
+  function onChange(evt: FormEvent<HTMLInputElement>) {
+
+    const dataTransferObject = new DataTransfer();
+
+    // get the previous files uploaded
+    if (files) {
+      forEach((f) => dataTransferObject.items.add(f), Array.from(files));
+    }
+
+    const inputFiles = evt.currentTarget.files,
+        normalizedFiles = !inputFiles?.length ? null : inputFiles;
+
+    if (normalizedFiles) {
+      forEach((f) => dataTransferObject.items.add(f), Array.from(normalizedFiles));
+    }
+
+    if (inputRef.current) {
+      inputRef.current.files = dataTransferObject.files;
+      onChangeProp(inputRef.current.files);
+    }
+  }
+
+  function openPicker() {
+    inputRef.current?.click();
+  }
+
+  function onDismiss(fileObj : File, selectedFile : HTMLElement | null) {
+    // if the selected file is either the last file in the container or the last remaining file, set focus accordingly
+    const closeBtns = Array.from(selectedFilesContainerRef.current?.
+        querySelectorAll<HTMLButtonElement>('.nx-selected-file__dismiss-btn') ?? []);
+    forEach((btn)=> {
+      if (selectedFile?.contains(btn) && btn === last(closeBtns)) {
+        closeBtns.length <= 1 ? inputRef.current?.focus() : closeBtns[closeBtns.length - 2].focus();
+      }
+    }, closeBtns);
+
+    const dataTransferObject = new DataTransfer();
+    if (inputRef.current) {
+      if (files) {
+        forEach((f: File) => dataTransferObject.items.add(f), without([fileObj], Array.from(files)));
+        inputRef.current.files = dataTransferObject.files;
+
+        const normalizedFiles = !inputRef.current.files.length ? null : inputRef.current.files;
+        onChangeProp(normalizedFiles);
+      }
+    }
+  }
+
+  // the container scrolls to the bottommost file only when new files are added
+  useEffect(function() {
+    const newFilesAdded = previousFiles === null || (files && (files.length > previousFiles.length));
```

Comment:
I haven't figured out a cleaner way of implementing this without using state. Design would like the container holding the selected files to scroll to the bottom when selected files are added, but not when removing a file.

Reply:
> I haven't figured out a cleaner way of implementing this without using state. Design would like the container holding the selected files to scroll to the bottom when selected files are added, but not when removing a file.

Would it work to simply add a `useEffect` to `SelectedFile` that scrolls it into view when it is first mounted?

Reply:
> Would it work to simply add a `useEffect` to `SelectedFile` that scrolls it into view when it is first mounted?
Great idea! Implemented here: [79ee526](https://github.com/sonatype/sonatype-react-shared-components/pull/850/commits/79ee5268a26abdc7919de896ca9eb9e9fcf21797)